### ISSUE TITLE
Route 53: throttling: status code 400

### ISSUE THREAD

**Issue by cmsmith7** [StackExchange/dnscontrol#1122]
Hi, we have been using dnscontrol for almost a year to manage 5400+ domains and think this tool is seriously awesome so it pains me to raise an issue we have suddenly started experiencing on a daily basis.  We manage these domains via a github repo and a set of github actions, we filter for the files changed in PR's/merges to main as to only update those hosted zones to limit api calls and also update them in a sequential order as opposed to concurrently to also further reduce api calls.  We have suddenly been hit with multiple throttling failures whereby we are being throttled where only a couple of domains have been modified with a new single entry:

Throttling: 
	status code: 400, request id: 707aaaaaaa-6eaa-40aa-95aa-a106aaaaaaaaa

As a side note we were running version 3.5.0 for the last few months, I upgraded this process to use 3.8.0 today and are still facing the issue.  AWS have a hard limit set in route 53 where they allow 5 api calls per second and this is non configurable.  Would it be possible to implement rate limiting/back off to ensure dnscontrol stays inside the limit mentioned above?

Many thanks.

**Comment by tlimoncelli** [comment]
5400 domains? That may be a new record!  I'm glad you find dnscontrol useful!

The Route53 provider has some rate-limiting code in it.  I didn't write it, so I'm not super familiar with it. That said, if it isn't working for you, that's a bug. 

Can you help me reproduce the problem?

I'm surprised that AWS has a 5 api call per second limit.   Are you sure that isn't configurable?  Can you check with your AWS rep to see if they have any suggestions?   We use R53 for our stackoverflow.com domain and haven't hit such limits.

I'm surprised that AWS gives a 400 code when throttling. Most give a `429 Too Many Requests` result. Could there be an underlying issue that is masked as a throttling issue?

**Comment by cmsmith7** [comment]
Many thanks for your quick response, unfortunately there is no way to consistently reproduce it just suddenly happens but has just been happening more often. We have AWS premier support and they have come back stating that it is a hard limit and cannot be changed (this is not the case, they just really do not like doing it as they say modifying this limit potentially unbalances their platform and they keep that limit set as it ensures QoS for all customers) I have had this changed for one other account (EKS utilising widely used add-on external-dns - AWS actually raised PR's against Kubernetes github repo external-dns on the back of that ticket to reduce API calls) whereby we have had that "api calls per second" limit increased from 5 to 10 and will go through the process of trying to get his account limit raised too.  

I have seen the issue whereby we have had multiple github PR's raised around the same time resulting in multiple github actions running/overlapping which can cause throttling which is completely understandable (will most likely put a locking mechanism in place to mitigate this) but I have also seen a single action run sequentially through multiple zones updating a single record that results in throttling.  The github update action performs a series of steps (omitting the setup):

- generate list of hosted zones in r53
- files changed - ascertain zones that have been updated in this PR (so we do not perform the process against all 5400+ zones for every PR), create files-changed.
- zone check - loop through zones in files-changed performing a dnscontrol check
- zone preview - loop through zones performing a dnscontrol preview - if zone exists push to (update-list) if not push to (create-zone list)
- zone create - creates new zones in (create-zone list) that do not exist - then push zone name in to the (update-list)
- zone push - loops through update-list performing a dnscontrol push

The throttling can occur during the preview or push step.  We have also asked AWS to advise/analyse if they can see any other possible limits/restrictions that could be causing the issue and am currently awaiting a response for that.

Many thanks again for your response.

**Comment by tlimoncelli** [comment]
More questions:

1. Does the problem happen only in certain steps?
2. Does the problem happen if you just run `dnscontrol push`?  (Side note: I'm personally curious how long a full run takes vs. your optimized version if you can share that.)
3. Does the "zone preview" step run dnscontrol once for each domain or does it list multiple domains
   - `dnscontrol preview --domains foo.com ; dnscontrol preview --domains bar.com ; dnscontrol preview --domains example.com`
   - vs.
   - `dnscontrol preview --domains foo.com,bar.com,example.com`
4. Have you tried modifying the Go code to add a 1/5th second pause after each API call?

**Comment by cmsmith7** [comment]
Absolutely ..

1. Yes the preview and push steps.
2. I don't tend to use dnscontrol from CLI much now apart from when testing new versions of the binary, the build stage of this job when converting/importing bind zone files and testing the github actions so haven't experienced it when running manually.  Sorry by full run do you mean push all zones at once (we never do this - due to the amount of zones), so we have only ever pushed/created zones that are updated in PR's since their migration from Bind to Route53, all updates done by code [except for when we require weighted entries and use IGNORE_NAME so we can manage the weights manually].  If you are referring to the length of time the github action takes to run, for a single record in a single zone approximately 1m24s with the preview taking 12s and the push taking 13s, for each additional zone preview it will take an additional 13s and for each additional zone push it takes an additional 14s.
3. The individual steps check, preview and push follow a similar logic, the zones are checked/previewed/pushed individually using a simple "dnscontrol [check,preview or push]" with a little logic after each zone push to confirm that the command was successful, move the zone into the correct list i.e create-zone or zone-push etc.
4. No and that is a very good idea, I have never coded in Go but will definitely look into that.

**Comment by tlimoncelli** [comment]
I'm unable to reproduce this.  Anything you can do to help create a reproduction case would be helpful.

Some random thoughts:

**Is this a problem with dnscontrol or the way dnscontrol is being called?**

To separate out those two possibilities, could you generate one big dnsconfig.js with all 5400+ domains and run "dnscontrol preview"?  If that doesn't get throttled, that would tell us something.

**What exactly is getting throttled?**

There is re-try code all over the place.  Maybe we're missing it on one place.  Can you determine where the `Throttling: status code: 400` message is being generated?

**What is getting throttled? (part 2)**

The code outputs `============ Route53 rate limit exceeded. Waiting` any time AWS tells it was throttled and it needs to back off.  Are you seeing those messages?

**Does the problem go away if we artificially slow dnscontrol?**

The problem happens on "preview" and "push".  The one thing that those have in common is calls to download the list of records for a zone: `fetchRecordSets()`

Try adding `time.Sleep(200 * time.Millisecond)` to the loop in `fetchRecordSets()`.  (Right after the `for {`).  This will make the system run much slower but the result will help us narrow down the problem.

**Comment by tlimoncelli** [comment]
FYI:

I made an attempt to reproduce this and sadly couldn't.

I tried running many "previews" at the same time. My hope was that they would throttle.  I realize that in your environment they aren't run in parallel, but I thought I'd try to make the test as difficult as possible.

I ran a number of "dnscontrol preview" in parallel and I didn't see this error.

I put a list of domains that are on Route53 in `/tmp/doms.txt` and ran this command:

```
for i in $(</tmp/doms.txt) ; do echo ========== $i ; ( dnscontrol preview --domains $i --providers r53 | awk '{ print }' &)  ; done
```

* The for loop loops over each line of `/tmp/doms.txt` (27 domains in that file)
* The `( ... &)` puts the preview command in the background so they run in parallel.
* The `awk '{ print }' ` just makes output pretty. It collects the output and outputs one line at a time.

The output I get shows no throttling and no errors.

I also tried removing the `&` to let them run individually.  No change.

Next I tried duplicating the /tmp/doms.txt file many times, which got errors when there was 270 processes running at once.  However those errors were:

```
Get "https://dns.googleapis.com/dns/v1/projects/stack-sre/managedZones?alt=json&pageToken=&prettyPrint=false": oauth2: cannot fetch token: Post "https://accounts.google.com/o/oauth2/token": net/http: TLS handshake timeout
```

I think that was the OAUTH2 system rate limiting, not the DNS API itself.

Lastly, I made a smaller test (81 domains) I did see it throttling, but the code backed off and everything completed fine.  The messages looked like:

```
Done. 0 corrections.
Throttling: 
	status code: 400, request id: 0a5b2850-fdfd-47d1-8f35-1e175ba2abe8
Throttling: 
	status code: 400, request id: 97d3cd94-ac0b-4db2-8ca1-fd7d0dd1dc46
Throttling: 
	status code: 400, request id: 4acf7882-2072-47d4-9698-81f724406d0e
Throttling: 
	status code: 400, request id: 3fa016b9-e514-4349-930d-ec2aac2aa121
Throttling: 
	status code: 400, request id: 264c7f1d-e3ef-4bdd-825f-c4cd45f83842
```

I see those messages but the re-tries happen as expected and the system completes.

Welp, at least I tried.  Maybe you'll have better luck reproducing the issue!

I hate to contradict the AWS support person you talked with, but I question the 5/sec rate limit.

P.S.  When you get the Throttling message, does the change not get made?  Does the program end with an error message or failure?

**Comment by cmsmith7** [comment]
Hi .. I am so happy to report back that dnscontrol was not the cause of this issue so I can happily retract my first line of my initial post, I suspected due to the extensive testing/discussions and all things taken into account i.e intermittency, updates of a single address that there was something else running in the account, so we reached out to AWS to delve deeper who in turn reached out to their internal team to to check timeline of errors and correlate those against our updates, it turns out we had a new security tool running scans which at points were exceeding the API call limit by as much as 40x. I can't thank you enough for all your time and for all the testing and responding to me on this issue .. I referred AWS to this issue when they were investigating so it would have given them confidence it was not a client side issue with dnscontrol. Many thanks again for all the help and for dnscontrol itself.
